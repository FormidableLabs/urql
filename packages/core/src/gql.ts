/* eslint-disable prefer-rest-params */
import { TypedDocumentNode } from '@graphql-typed-document-node/core';

import {
  DocumentNode,
  DefinitionNode,
  FragmentDefinitionNode,
  Kind,
} from 'graphql';

import { keyDocument, stringifyDocument } from './utils';

const applyDefinitions = (
  fragmentNames: Map<string, string>,
  target: DefinitionNode[],
  source: Array<DefinitionNode> | ReadonlyArray<DefinitionNode>
) => {
  for (let i = 0; i < source.length; i++) {
    const definition = source[i];
    if (source[i].kind === Kind.FRAGMENT_DEFINITION) {
      const name = (source[i] as FragmentDefinitionNode).name.value;
      const value = stringifyDocument(definition);
      // Fragments will be deduplicated according to this Map
      const prevValue = fragmentNames.get(name);
      if (prevValue === undefined) {
        fragmentNames.set(name, value);
        target.push(source[i]);
      } else if (process.env.NODE_ENV !== 'production' && prevValue !== value) {
        // Fragments with the same names is expected to have the same contents
        console.warn(
          '[WARNING: Duplicate Fragment] A fragment with name `' +
            name +
            '` already exists in this document.\n' +
            'While fragment names may not be unique across your source, each name must be unique per document.'
        );
      }
    } else {
      target.push(source[i]);
    }
  }
};

function gql<Data = any, Variables = object>(
  strings: TemplateStringsArray,
  ...interpolations: Array<TypedDocumentNode | DocumentNode | string>
): TypedDocumentNode<Data, Variables>;

function gql<Data = any, Variables = object>(
  string: string
): TypedDocumentNode<Data, Variables>;

function gql(/* arguments */) {
  const fragmentNames = new Map<string, string>();
  const definitions: DefinitionNode[] = [];
  const interpolations: DocumentNode[] = [];

  // Apply the entire tagged template body's definitions
  let body: string = Array.isArray(arguments[0])
    ? arguments[0][0]
    : arguments[0] || '';
  for (let i = 1; i < arguments.length; i++) {
    const value = arguments[i];
    if (value && value.kind === Kind.DOCUMENT) {
      interpolations.push(value);
    } else {
      body += value;
    }

    body += arguments[0][i];
  }

  // Apply the tag's body definitions
  applyDefinitions(fragmentNames, definitions, keyDocument(body).definitions);

  // Copy over each interpolated document's definitions
  for (let i = 0; i < interpolations.length; i++)
    applyDefinitions(fragmentNames, definitions, interpolations[i].definitions);

  return keyDocument({
    kind: Kind.DOCUMENT,
    definitions,
  });
}

export { gql };
