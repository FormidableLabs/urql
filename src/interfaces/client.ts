import Observable from 'zen-observable-ts';

import { IClientEvent } from '../modules/events';
import { ICache } from './cache';
import { IExchangeResult, IExecutionData } from './exchange';
import { IQuery } from './query';

export interface IClient {
  cache: ICache;
  executeSubscription$(subscriptionObject: IQuery): Observable<IExchangeResult>;
  executeQuery$(
    queryObject: IQuery,
    skipCache: boolean
  ): Observable<IExchangeResult>;
  executeQuery(
    queryObject: IQuery,
    skipCache: boolean
  ): Promise<IExchangeResult>;
  executeMutation$(mutationObject: IQuery): Observable<IExecutionData>;
  executeMutation(mutationObject: IQuery): Promise<IExecutionData>;
  invalidateQuery(queryObject: IQuery): Promise<void>;
  refreshAllFromCache(): void;
  subscribe(callback: (event: IClientEvent) => void): () => void;
}
