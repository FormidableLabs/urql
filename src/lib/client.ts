import {
  filter,
  makeSubject,
  pipe,
  publish,
  share,
  Source,
  Subject,
  subscribe,
  take,
} from 'wonka';

import { defaultExchanges } from '../exchanges';
import { CombinedError } from './error';
import { hashString } from './hash';

import {
  Exchange,
  ExchangeResult,
  GraphQLRequest,
  Mutation,
  Operation,
  OperationContext,
  OperationType,
  Query,
  Subscription,
} from '../types';

/** Options for configuring the URQL [client]{@link Client}. */
export interface ClientOptions {
  /** Target endpoint URL such as `https://my-target:8080/graphql`. */
  url: string;
  /** Any additional options to pass to fetch. */
  fetchOptions?: RequestInit | (() => RequestInit);
  /** An ordered array of Exchanges. */
  exchanges?: Exchange[];
}

/** The URQL application-wide client library. Each execute method starts a GraphQL request and
 returns a stream of results. */
export class Client implements ClientOptions {
  // These are variables derived from ClientOptions
  url: string;
  fetchOptions: RequestInit;
  exchanges: Exchange[];

  // These are internals to be used to keep track of operations
  dispatchOperation: (operation: Operation) => void;
  operations$: Source<Operation>;
  results$: Source<ExchangeResult>;

  constructor(opts: ClientOptions) {
    this.url = opts.url;

    this.fetchOptions =
      typeof opts.fetchOptions === 'function'
        ? opts.fetchOptions()
        : (opts.fetchOptions || {});

    this.exchanges =
      opts.exchanges !== undefined
        ? opts.exchanges
        : defaultExchanges;

    // This subject forms the input of operations; executeOperation may be
    // called to dispatch a new operation on the subject
    const [operations$, nextOperation] = makeSubject<Operation>();
    this.operations$ = operations$;
    this.dispatchOperation = nextOperation;

    // All operations run through the exchanges in a pipeline-like fashion
    // and this observable then combines all their results
    this.results$ = pipeExchanges(this, this.exchanges, this.operations$);
  }

  createOperation = (
    type: OperationType,
    query: GraphQLRequest,
    context: Partial<OperationContext>
  ): Operation => ({
    ...query,
    key: hashString(JSON.stringify(query)),
    operationName: type,
    context: {
      url: this.url,
      fetchOptions: this.fetchOptions,
      ...context
    }
  });

  executeOperation(operation: Operation): Source<ExchangeResult> {
    this.dispatchOperation(operation);
    const isSpecificExchangeResult = filter((res: ExchangeResult) => res.operation.key === operation.key);

    if (operation.operationName === 'mutation') {
      return pipe(this.results$, isSpecificExchangeResult, take(1));
    } else {
      return pipe(this.results$, isSpecificExchangeResult);
    }
  }

  executeQuery = (query: Query, opts: Partial<OperationContext>): Source<ExchangeResult> =>
    this.executeOperation(this.createOperation('query', query, opts));

  executeSubscription = (query: Subscription, opts: Partial<OperationContext>): Source<ExchangeResult> =>
    this.executeOperation(this.createOperation('subscription', query, opts));

  executeMutation = (query: Mutation, opts: Partial<OperationContext>): Source<ExchangeResult> =>
    this.executeOperation(this.createOperation('mutation', query, opts));
}

/** Create pipe of exchanges */
const pipeExchanges = (
  client: Client,
  exchanges: Exchange[],
  operation$: Source<Operation>
): Source<ExchangeResult> => {
  /** Recursively pipe to each exchange */
  const callExchanges = (value: Source<Operation>, index: number = 0) => {
    if (index >= exchanges.length) {
      return value;
    }

    const currentExchange = exchanges[index];
    return currentExchange({
      forward: val => callExchanges(val, index + 1),
      client,
    })(value);
  };

  return share(callExchanges(operation$, 0));
};
