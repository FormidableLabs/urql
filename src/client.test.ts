import { print } from 'graphql';
import gql from 'graphql-tag';

/** NOTE: Testing in this file is designed to test both the client and it's interaction with default Exchanges */

jest.mock('./utils/keyForQuery', () => ({
  getKeyForQuery: () => 123,
  getKeyForRequest: () => 123,
}));

import { map, pipe, subscribe, tap, interval } from 'wonka';
import { createClient } from './client';

const url = 'https://hostname.com';

describe('createClient', () => {
  it('passes snapshot', () => {
    const c = createClient({
      url,
    });

    expect(c).toMatchSnapshot();
  });
});

const query = {
  key: 1,
  query: gql`
    {
      todos {
        id
      }
    }
  `,
  variables: { example: 1234 },
};

let receivedOps: any[] = [];
let client = createClient({ url: '1234' });
const receiveMock = jest.fn(s =>
  pipe(
    s,
    tap(op => (receivedOps = [...receivedOps, op])),
    map(op => ({ operation: op }))
  )
);
const exchangeMock = jest.fn(() => receiveMock);

beforeEach(() => {
  receivedOps = [];
  exchangeMock.mockClear();
  receiveMock.mockClear();
  client = createClient({ url, exchanges: [exchangeMock] as any[] });
});

describe('exchange args', () => {
  it('receives forward function', () => {
    // @ts-ignore
    expect(typeof exchangeMock.mock.calls[0][0].forward).toBe('function');
  });

  it('recieves client', () => {
    // @ts-ignore
    expect(exchangeMock.mock.calls[0][0]).toHaveProperty('client', client);
  });
});

describe('executeQuery', () => {
  it('passes query string exchange', () => {
    pipe(
      client.executeQuery(query),
      subscribe(x => x)
    );

    const receivedQuery = receivedOps[0].query;
    expect(print(receivedQuery)).toBe(print(query.query));
  });

  it('passes variables type to exchange', () => {
    pipe(
      client.executeQuery(query),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('variables', query.variables);
  });

  it('passes operationName type to exchange', () => {
    pipe(
      client.executeQuery(query),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('operationName', 'query');
  });

  it('passes url (from context) to exchange', () => {
    pipe(
      client.executeQuery(query),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('context.url', url);
  });
});

describe('executeMutation', () => {
  it('passes query string exchange', async () => {
    pipe(
      client.executeMutation(query),
      subscribe(x => x)
    );

    const receivedQuery = receivedOps[0].query;
    expect(print(receivedQuery)).toBe(print(query.query));
  });

  it('passes variables type to exchange', () => {
    pipe(
      client.executeMutation(query),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('variables', query.variables);
  });

  it('passes operationName type to exchange', () => {
    pipe(
      client.executeMutation(query),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('operationName', 'mutation');
  });

  it('passes url (from context) to exchange', () => {
    pipe(
      client.executeMutation(query),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('context.url', url);
  });
});

describe('executeSubscription', () => {
  it('passes query string exchange', async () => {
    pipe(
      client.executeSubscription(query),
      subscribe(x => x)
    );

    const receivedQuery = receivedOps[0].query;
    expect(print(receivedQuery)).toBe(print(query.query));
  });

  it('passes variables type to exchange', () => {
    pipe(
      client.executeSubscription(query),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('variables', query.variables);
  });

  it('passes operationName type to exchange', () => {
    pipe(
      client.executeSubscription(query),
      subscribe(x => x)
    );

    expect(receivedOps[0]).toHaveProperty('operationName', 'subscription');
  });
});

describe('createQuery', () => {
  const queryString = query.query;

  beforeEach(() => {
    // @ts-ignore
    client.executeQuery = jest.fn(() =>
      pipe(
        interval(400),
        map((i: number) => ({ data: i, error: i + 1 }))
      )
    );
  });

  afterAll(() => {
    //@ts-ignore
    client.executeQuery.mockClear();
  });

  it('should return a query function', () => {
    const myQuery = client.createQuery({ query: queryString });
    expect(myQuery().then).toBeInstanceOf(Function);
  });

  it('should call executeQuery with the initial variables', async () => {
    const myQuery = client.createQuery({
      query: queryString,
      variables: { foo: 'bar' },
    });

    await myQuery();
    //@ts-ignore
    expect(client.executeQuery.mock.calls[0][0].variables.foo).toBe('bar');
  });

  it('should call executeQuery with the new variables', async () => {
    const myQuery = client.createQuery({
      query: queryString,
      variables: { foo: 'bar' },
    });

    await myQuery({ foo: 'baz' });
    // @ts-ignore
    expect(client.executeQuery.mock.calls[0][0].variables.foo).toBe('baz');
  });
});

describe('createMutation', () => {
  const queryString = query.query;

  beforeEach(() => {
    // @ts-ignore
    client.executeMutation = jest.fn(() =>
      pipe(
        interval(400),
        map((i: number) => ({ data: i, error: i + 1 }))
      )
    );
  });

  afterAll(() => {
    //@ts-ignore
    client.executeMutation.mockClear();
  });

  it('should return a query function', () => {
    const myMutation = client.createMutation(queryString);
    expect(myMutation().then).toBeInstanceOf(Function);
  });

  it('should call executeQuery with the supplied variables', async () => {
    const myMutation = client.createMutation(queryString);

    await myMutation({ foo: 'bar' });
    //@ts-ignore
    expect(client.executeMutation.mock.calls[0][0].variables.foo).toBe('bar');
  });
});
