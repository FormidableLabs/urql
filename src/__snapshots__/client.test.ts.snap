// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createClient passes snapshot 1`] = `
Client {
  "activeOperations": Object {},
  "createMutation": [Function],
  "createOperationContext": [Function],
  "createQuery": [Function],
  "createRequestOperation": [Function],
  "dispatchOperation": [Function],
  "exchange": [Function],
  "executeMutation": [Function],
  "executeQuery": [Function],
  "executeSubscription": [Function],
  "fetchOptions": undefined,
  "operations$": [Function],
  "reexecuteOperation": [Function],
  "results$": [Function],
  "url": "https://hostname.com",
}
`;
